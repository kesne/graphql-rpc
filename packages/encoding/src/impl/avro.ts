import avro from 'avsc';
import { parse, GraphQLSchema, OperationDefinitionNode, FieldNode } from 'graphql';

export function encode(schema: GraphQLSchema, query: string) {
    const Request = avro.Type.forSchema({
        name: 'Request',
        type: 'record',
        fields: [
            {
                name: 'type',
                type: ['null', { name: 'Type', type: 'enum', symbols: ['QUERY', 'MUTATION', 'SUBSCRIPTION'] }],
                default: null
            },
            { name: 'field', type: ['null', 'int'], default: null },
            { name: 'fields', type: 'long' },
            { name: 'selections', type: ['null', { type: 'array', items: ['Request'] }], default: null }
        ]
    });

    const requestDescription: any = {
        fields: 0
    };

    const parsedQuery = parse(query);

    const operationDefinition = parsedQuery.definitions[0] as OperationDefinitionNode;

    switch (operationDefinition.operation) {
        case 'mutation':
            requestDescription.type = 'MUTATION';
            break;
        case 'subscription':
            requestDescription.type = 'SUBSCRIPTION';
        case 'query':
        default:
            // requestDescription.type = 'QUERY';
            break;
    }

    function walkSelections(selectionDescription: any, node: FieldNode, schemaNode: any) {
        const selections = node.selectionSet!.selections;
        const schemaFields = schemaNode.getFields();

        selections.forEach(selection => {
            if (selection.kind !== 'Field') {
                throw new Error('Only field selections are currently supported.');
            }

            const selectionNode = schemaFields[selection.name.value];
            const fieldDirective = selectionNode.astNode.directives.find(
                (dir: any) => dir.name.value === 'field'
            );
            // Only ever has one argument right now:
            const fieldNumber = parseInt(fieldDirective.arguments[0].value.value, 10) - 1;

            selectionDescription.fields |= 1 << fieldNumber;

            if (selection.selectionSet) {
                if (!selectionDescription.selections) {
                    selectionDescription.selections = [];
                }

                const data = { field: fieldNumber, fields: 0 };
                selectionDescription.selections.push(data);

                walkSelections(
                    data,
                    selection,
                    resolveType(schemaNode.getFields()[selection.name.value])
                );
            }
        });
    }

    function resolveType(node: any): any {
        if (node.ofType) {
            return resolveType(node.ofType);
        }

        if (node.type && node.type.ofType) {
            return resolveType(node.type.ofType);
        }

        if (node.type) {
            return resolveType(node.type);
        }

        return node;
    }

    walkSelections(requestDescription, operationDefinition as any, schema.getQueryType());

    return Request.toBuffer(requestDescription);
}
