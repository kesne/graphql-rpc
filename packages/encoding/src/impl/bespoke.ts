import varint from 'varint';
import { GraphQLSchema } from 'graphql';
import createIR, { IR, ChildIR } from '../createIR';
import { Variables } from '../types';

// NOTE: Even though we never use 0, we still reserve it and don't use it for mutations to make it side-effect-free to zero-fill the buffer.
const TYPE_QUERY = 0;
const TYPE_MUTATION = 1;
const TYPE_SUBSCRIPTION = 2;
const FIELD_EXIT = 3;
const FIELD_ARGUMENT = 4;
// Use an enter offset to make it unambiguous with query (0) mutation (1) subscription (2) exit (3) and arguments (4)
const FIELD_ENTER_OFFSET = 6;

function stripTrailingExits(data: number[]) {
    let i = data.length - 1;
    while (data[i] === FIELD_EXIT) {
        data.splice(i, 1);
        i--;
    }
}

export function encode(schema: GraphQLSchema, query: string, variables: Variables) {
    const data: number[] = [];

    function write(num: number | number[]) {
        if (Array.isArray(num)) {
            num.forEach(n => write(n));
        } else {
            varint.encode(num).forEach(byte => {
                data.push(byte);
            });
        }
    }

    function walkIR(current: ChildIR | IR) {
        write(current.fields);
        if ('arguments' in current && current.arguments) {
            write(FIELD_ARGUMENT);
            for (const argument of current.arguments) {
                write(argument.arg);
                // TODO: Really should get better null handling:
                write(argument.value ? argument.value.length : 0);
                if (argument.value) {
                    write([...argument.value]);
                }
            }
        }

        if (current.selections) {
            for (const selection of current.selections) {
                write(FIELD_ENTER_OFFSET + selection.id);
                walkIR(selection);
                write(FIELD_EXIT);
            }
        }
    }

    const ir = createIR(schema, query, variables, {
        getType: type => type
    });

    walkIR(ir);
    stripTrailingExits(data);

    switch (ir.type) {
        case 'mutation':
            write(TYPE_MUTATION);
            break;
        case 'subscription':
            write(TYPE_SUBSCRIPTION);
            break;
    }

    return Buffer.from(data);
}

// NOTE: This isn't complete, it's just a proof-of-concept:
function decode(data: number[]) {
    const root: any = {};
    let offset = 0;

    function resolveField(selection = root) {
        selection.fields = varint.decode(data, offset);
        offset += varint.decode.bytes;

        if (data[offset]) {
            // We're pushing a field:
            if (data[offset] >= FIELD_ENTER_OFFSET) {
                selection.children = selection.children || {};
                const field = varint.decode(data, offset) - FIELD_ENTER_OFFSET;
                offset += varint.decode.bytes;
                selection.children[field] = {};
                resolveField(selection.children[field]);
            }
        }
    }

    resolveField();

    return root;
}
