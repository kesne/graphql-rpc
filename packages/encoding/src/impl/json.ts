import { GraphQLSchema } from 'graphql';
import compress from 'graphql-query-compress';

export function encode(schema: GraphQLSchema, query: string) {
    return JSON.stringify({
        query: compress(query)
    });
}

export function decode(schema: GraphQLSchema, blob: Buffer) {
    return JSON.parse(blob.toString('utf-8'));
}
