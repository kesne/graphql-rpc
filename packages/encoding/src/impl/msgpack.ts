import msgpack from 'msgpack5';
import { GraphQLSchema } from 'graphql';
import createIR, { IR } from '../createIR';
import { Variables } from '../types';

function optimizeIR(ir: IR) {
    let optimized: any = ir;

    if (ir.selections) {
        ir.selections.forEach(selection => optimizeIR(selection));
    }

    if (ir.arguments) {
        ir.arguments.forEach(arg => optimizeIR(arg));
    }

    Object.keys(ir).forEach(key => {
        if (key.length > 1) {
            optimized[key.charAt(0)] = optimized[key];
            delete optimized[key];
        }
    });
}

export function encode(schema: GraphQLSchema, query: string, variables: Variables) {
    const encoder = msgpack();

    const ir = createIR(schema, query, variables, {
        getType(type) {
            if (type === 'mutation') return 1;
            if (type === 'subscription') return 2;
        }
    });

    optimizeIR(ir);

    return Buffer.from(encoder.encode(ir).toString('hex'), 'hex');
}
