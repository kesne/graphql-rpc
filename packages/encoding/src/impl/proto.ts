import path from 'path';
import protobuf from 'protobufjs';
import { GraphQLSchema } from 'graphql';
import createIR from '../createIR';
import { Variables } from '../types';

export function encode(schema: GraphQLSchema, query: string, variables: Variables) {
    const root = protobuf.loadSync(path.join(__dirname, 'request.proto'));
    const Request = root.lookupType('RequestSelection');

    const ir = createIR(schema, query, variables, {
        getType(type) {
            if (type === 'mutation') return 1;
            if (type === 'subscription') return 2;
        }
    });

    return Request.encode(Request.fromObject(ir)).finish();
}
