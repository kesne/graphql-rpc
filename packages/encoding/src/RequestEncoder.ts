import { GraphQLSchema } from "graphql";
import * as impl from './impl/json';

export default class RequestEncoder {
    constructor(private schema: GraphQLSchema, private implementation = impl) {}

    encode(query: string) {
        return this.implementation.encode(this.schema, query);
    }
}
