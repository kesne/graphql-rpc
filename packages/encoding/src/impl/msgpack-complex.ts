import msgpack from 'msgpack5';
import { parse, GraphQLSchema, OperationDefinitionNode, FieldNode, GraphQLField } from 'graphql';

export function encode(schema: GraphQLSchema, query: string) {
    const parsedQuery = parse(query);
    const encoder = msgpack();

    const payload: any = {};

    const operationDefinition = parsedQuery.definitions[0] as OperationDefinitionNode;

    switch (operationDefinition.operation) {
        case 'mutation':
            payload.t = 1;
            break;
        case 'subscription':
            payload.t = 2;
        case 'query':
        default:
            // Do nothing, default is query:
            break;
    }

    payload.s = { f: 0 };

    function walkSelections(s: any, node: FieldNode, schemaNode: any) {
        const selections = node.selectionSet!.selections;

        // console.log(schemaNode);
        const schemaFields = schemaNode.getFields();

        selections.forEach((selection, i) => {
            if (selection.kind !== 'Field') {
                throw new Error('Only field selections are currently supported.');
            }

            const selectionNode = schemaFields[selection.name.value];
            const fieldDirective = selectionNode.astNode.directives.find((dir: any) => dir.name.value === 'field');
            // @ts-ignore Only has one argument right now:
            const fieldNumber = parseInt(fieldDirective!.arguments[0].value.value, 10) - 1;

            s.f |= (1 << fieldNumber);

            if (selection.selectionSet) {
                if (!s.s) {
                    s.s = {};
                }

                if (!s.s[fieldNumber]) {
                    s.s[fieldNumber] = { f: 0 };
                }

                walkSelections(s.s[fieldNumber], selection, resolveType(schemaNode.getFields()[selection.name.value]));
            }
        });
    }

    function resolveType(node: any): any {
        if (node.ofType) {
            return resolveType(node.ofType);
        }

        if (node.type && node.type.ofType) {
            return resolveType(node.type.ofType);
        }

        if (node.type) {
            return resolveType(node.type);
        }

        return node;
    }

    walkSelections(payload.s, operationDefinition as any, schema.getQueryType());

    return Buffer.from(
        encoder
            .encode(payload)
            .toString('hex'),
        'hex'
    );
}
