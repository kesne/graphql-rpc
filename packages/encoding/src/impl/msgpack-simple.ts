import msgpack from 'msgpack5';
import { GraphQLSchema } from 'graphql';
import compress from 'graphql-query-compress';

export function encode(schema: GraphQLSchema, query: string) {
    const encoder = msgpack();
    return Buffer.from(
        encoder
            .encode({
                query: compress(query)
            })
            .toString('hex'),
        'hex'
    );
}
