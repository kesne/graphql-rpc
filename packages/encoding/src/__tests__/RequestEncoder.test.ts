import fs from 'fs';
import path from 'path';
import zlib from 'zlib';
import fastmatter from 'fastmatter';
import { buildSchema } from 'graphql';
import RequestEncoder from '../RequestEncoder';
import * as baselineImpl from '../impl/json';

const FIXTURES = ['minimal'];

function getPercentageChange(oldNumber: number, newNumber: number) {
    var decreaseValue = oldNumber - newNumber;

    return ((decreaseValue / oldNumber) * 100).toFixed(2) + '%';
}

describe('encoding', () => {
    FIXTURES.forEach(fixtureSet => {
        const schema = buildSchema(
            fs.readFileSync(path.join(__dirname, 'fixtures', fixtureSet, 'schema.graphql'), 'utf8')
        );

        const baselineEncoder = new RequestEncoder(schema, baselineImpl);
        const testEncoder = new RequestEncoder(schema);

        const files = fs.readdirSync(path.join(__dirname, 'fixtures', fixtureSet, 'queries'));

        describe(fixtureSet, () => {
            files.forEach(filename => {
                it(`encoding query "${filename}"`, () => {
                    const queryFile = fs.readFileSync(
                        path.join(__dirname, 'fixtures', fixtureSet, 'queries', filename),
                        'utf8'
                    );

                    const { body: query, attributes } = fastmatter(queryFile);

                    const baselineEncoding = baselineEncoder.encode(query, attributes.variables);
                    const testEncoding = testEncoder.encode(query, attributes.variables);

                    expect({
                        baselineSize: baselineEncoding.length,
                        size: testEncoding.length,
                        difference: getPercentageChange(
                            baselineEncoding.length,
                            testEncoding.length
                        ),
                        optimizedDifference: getPercentageChange(
                            Math.min(
                                zlib.gzipSync(baselineEncoding).length,
                                baselineEncoding.length
                            ),
                            Math.min(zlib.gzipSync(testEncoding).length, testEncoding.length)
                        )
                    }).toMatchSnapshot();
                });
            });
        });
    });
});
