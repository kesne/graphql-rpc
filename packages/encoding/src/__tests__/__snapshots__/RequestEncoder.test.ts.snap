// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`encoding minimal encoding query "all.graphql" 1`] = `
Object {
  "compressedDifference": "4.29%",
  "difference": "0.00%",
  "size": 67,
}
`;

exports[`encoding minimal encoding query "nested.graphql" 1`] = `
Object {
  "compressedDifference": "-192.86%",
  "difference": "0.00%",
  "size": 205,
}
`;

exports[`encoding minimal encoding query "one.graphql" 1`] = `
Object {
  "compressedDifference": "36.59%",
  "difference": "0.00%",
  "size": 26,
}
`;
